configuration/pih/liquibase/sql/create_functions.sql

```diff
@@ -2180,6 +2180,33 @@ order by ps.start_date desc limit 1;
 
     RETURN state_name_out;
 
+END
+#
+/* 
+The following accepts a patient id and program (identified by UUID or program id)
+it will return the the date of the latest time the patient enrolled in that program
+ */
+#
+DROP FUNCTION IF EXISTS latestProgramEnrollmentDate;
+#
+CREATE FUNCTION latestProgramEnrollmentDate(_patient_id int(11), _program_id_or_uuid char(38))
+    RETURNS datetime
+    DETERMINISTIC
+
+BEGIN
+
+  DECLARE ret datetime;
+
+select max(pp.date_enrolled) into ret
+from patient_program pp 
+inner join program p on p.program_id = pp.program_id
+where pp.patient_id  = _patient_id
+and (pp.program_id = _program_id_or_uuid or p.uuid = _program_id_or_uuid)
```

Comment:
This is kind of odd, but if it works...

Reply:
Based on how I'm seeing you use (or abuse) this in the other query, for safety I'd recommend more and more explicity functions:

1. get program_id given concept mapping
2. get workflow_id given _program_id and concept mapping
3. get state_id given _workflow_id and concept mapping
4. get latest program xyz given _patient_id and _program_id

Reply:
Did you make any of these suggested changes?

Reply:
There really isn't a need for this script to add these other functions.  You'll see in the main script, when getting the minimum patient_program_id, I already have the only database row I need.  Same with the patient_state.  We can discuss more if you think I should add additional functions for this.